File: Proyecto_Ordenamientos/Proyecto_Ordenamientos.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <chrono>
#include <fstream>
#include <random>
using namespace std;
using namespace std::chrono;

// ----------------- ALGORITMOS -----------------

void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n-1; i++)
        for (int j = 0; j < n-i-1; j++)
            if (arr[j] > arr[j+1])
                swap(arr[j], arr[j+1]);
}

void insertionSort(vector<int>& arr) {
    for (int i = 1; i < arr.size(); i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

void selectionSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = 0; i < n-1; i++) {
        int minIdx = i;
        for (int j = i+1; j < n; j++)
            if (arr[j] < arr[minIdx]) minIdx = j;
        swap(arr[i], arr[minIdx]);
    }
}

int partitionQS(vector<int>& arr, int low, int high) {
    int pivot = arr[high];
    int i = (low - 1);
    for (int j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            swap(arr[i], arr[j]);
        }
    }
    swap(arr[i + 1], arr[high]);
    return (i + 1);
}

void quickSort(vector<int>& arr, int low, int high) {
    if (low < high) {
        int pi = partitionQS(arr, low, high);
        quickSort(arr, low, pi - 1);
        quickSort(arr, pi + 1, high);
    }
}

void merge(vector<int>& arr, int l, int m, int r) {
    int n1 = m - l + 1, n2 = r - m;
    vector<int> L(n1), R(n2);
    for (int i = 0; i < n1; i++) L[i] = arr[l + i];
    for (int j = 0; j < n2; j++) R[j] = arr[m + 1 + j];
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2)
        arr[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
}

void mergeSort(vector<int>& arr, int l, int r) {
    if (l >= r) return;
    int m = l + (r - l) / 2;
    mergeSort(arr, l, m);
    mergeSort(arr, m + 1, r);
    merge(arr, l, m, r);
}

void heapify(vector<int>& arr, int n, int i) {
    int largest = i, l = 2*i+1, r = 2*i+2;
    if (l < n && arr[l] > arr[largest]) largest = l;
    if (r < n && arr[r] > arr[largest]) largest = r;
    if (largest != i) {
        swap(arr[i], arr[largest]);
        heapify(arr, n, largest);
    }
}

void heapSort(vector<int>& arr) {
    int n = arr.size();
    for (int i = n/2 - 1; i >= 0; i--)
        heapify(arr, n, i);
    for (int i = n-1; i > 0; i--) {
        swap(arr[0], arr[i]);
        heapify(arr, i, 0);
    }
}

// Hash sort (simplificado, usando hash directo)
void hashSort(vector<int>& arr) {
    int maxVal = *max_element(arr.begin(), arr.end());
    vector<int> hashTable(maxVal + 1, 0);
    for (int val : arr) hashTable[val]++;
    arr.clear();
    for (int i = 0; i <= maxVal; i++)
        for (int j = 0; j < hashTable[i]; j++)
            arr.push_back(i);
}

// Bucket Sort (para datos pequeños)
void bucketSort(vector<int>& arr) {
    int n = arr.size();
    int maxVal = *max_element(arr.begin(), arr.end());
    vector<int> bucket(maxVal + 1, 0);
    for (int val : arr) bucket[val]++;
    arr.clear();
    for (int i = 0; i <= maxVal; i++)
        for (int j = 0; j < bucket[i]; j++)
            arr.push_back(i);
}

// Radix Sort
int getMax(vector<int>& arr) {
    return *max_element(arr.begin(), arr.end());
}

void countSort(vector<int>& arr, int exp) {
    int n = arr.size();
    vector<int> output(n), count(10, 0);
    for (int i = 0; i < n; i++) count[(arr[i]/exp)%10]++;
    for (int i = 1; i < 10; i++) count[i] += count[i-1];
    for (int i = n - 1; i >= 0; i--) {
        output[count[(arr[i]/exp)%10] - 1] = arr[i];
        count[(arr[i]/exp)%10]--;
    }
    for (int i = 0; i < n; i++) arr[i] = output[i];
}

void radixSort(vector<int>& arr) {
    int m = getMax(arr);
    for (int exp = 1; m/exp > 0; exp *= 10)
        countSort(arr, exp);
}

// ----------------- UTILIDADES -----------------

void generarArreglo(vector<int>& arr, int n, int tipo) {
    arr.resize(n);
    random_device rd; mt19937 gen(rd());
    uniform_int_distribution<> dis(0, n);
    for (int i = 0; i < n; i++) arr[i] = dis(gen);
    if (tipo == 1) sort(arr.begin(), arr.end());
    else if (tipo == 2) reverse(arr.begin(), arr.end());
}

double medirTiempo(void (*func)(vector<int>&), vector<int> arr) {
    auto inicio = high_resolution_clock::now();
    func(arr);
    auto fin = high_resolution_clock::now();
    return duration<double, milli>(fin - inicio).count();
}

// ----------------- MAIN -----------------

int main() {
    vector<int> tamanos = {100, 1000, 10000};
    vector<string> tipos = {"Aleatorio", "Ordenado", "Inverso"};

    ofstream archivo("resultados.csv");
    archivo << "Algoritmo,Tamano,Tipo,Tiempo(ms)\n";

    vector<pair<string, void(*)(vector<int>&)>> algoritmos = {
        {"Bubble Sort", bubbleSort},
        {"Insertion Sort", insertionSort},
        {"Selection Sort", selectionSort},
        {"Quick Sort", [](vector<int>& a){quickSort(a,0,a.size()-1);}},
        {"Merge Sort", [](vector<int>& a){mergeSort(a,0,a.size()-1);}},
        {"Heap Sort", heapSort},
        {"Hash Sort", hashSort},
        {"Bucket Sort", bucketSort},
        {"Radix Sort", radixSort}
    };

    for (int n : tamanos) {
        for (int t = 0; t < 3; t++) {
            vector<int> base;
            generarArreglo(base, n, t);
            for (auto& alg : algoritmos) {
                double tiempo = medirTiempo(alg.second, base);
                archivo << alg.first << "," << n << "," << tipos[t] << "," << tiempo << "\n";
                cout << alg.first << " (" << tipos[t] << ", n=" << n << "): " << tiempo << " ms\n";
            }
        }
    }

    archivo.close();
    cout << "\nResultados guardados en 'resultados.csv'.\n";
    return 0;
}

File: Proyecto_Ordenamientos/Proyecto_Ordenamientos.py
# ==============================================================
# Se prueban 9 algoritmos de ordenamiento.
# en 12 escenarios distintos.
# Se mide el tiempo de ejecución de cada uno y se grafican resultados.
# ==============================================================

import random
import time
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os

# 1. IMPLEMENTACIÓN DE LOS MÉTODOS DE ORDENAMIENTO

# --- Bubble Sort ---
def bubble_sort(arr):
    """Ordenamiento burbuja: compara pares adyacentes e intercambia si están desordenados."""
    n = len(arr)
    for i in range(n):
        for j in range(0, n - i - 1):
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]

# --- Insertion Sort ---
def insertion_sort(arr):
    """Inserta cada elemento en la posición correcta dentro de la parte ya ordenada."""
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key

# --- Selection Sort ---
def selection_sort(arr):
    """Busca el mínimo en el arreglo restante y lo coloca en su posición."""
    for i in range(len(arr)):
        min_idx = i
        for j in range(i + 1, len(arr)):
            if arr[j] < arr[min_idx]:
                min_idx = j
        arr[i], arr[min_idx] = arr[min_idx], arr[i]

# --- Quick Sort ---
def quick_sort(arr):
    """Ordenamiento rápido (Quick Sort) recursivo."""
    if len(arr) <= 1:
        return arr
    pivot = arr[len(arr) // 2]
    left = [x for x in arr if x < pivot]
    middle = [x for x in arr if x == pivot]
    right = [x for x in arr if x > pivot]
    return quick_sort(left) + middle + quick_sort(right)

# --- Merge Sort ---
def merge_sort(arr):
    """Divide el arreglo en mitades, las ordena y las combina."""
    if len(arr) > 1:
        mid = len(arr)//2
        L = arr[:mid]
        R = arr[mid:]
        merge_sort(L)
        merge_sort(R)
        i = j = k = 0
        # Combina las mitades ordenadas
        while i < len(L) and j < len(R):
            if L[i] < R[j]:
                arr[k] = L[i]
                i += 1
            else:
                arr[k] = R[j]
                j += 1
            k += 1
        # Agrega los elementos restantes
        while i < len(L):
            arr[k] = L[i]
            i += 1
            k += 1
        while j < len(R):
            arr[k] = R[j]
            j += 1
            k += 1

# --- Heap Sort ---
def heapify(arr, n, i):
    """Reorganiza un subárbol para mantener la propiedad del heap."""
    largest = i
    l = 2 * i + 1
    r = 2 * i + 2
    if l < n and arr[l] > arr[largest]:
        largest = l
    if r < n and arr[r] > arr[largest]:
        largest = r
    if largest != i:
        arr[i], arr[largest] = arr[largest], arr[i]
        heapify(arr, n, largest)

def heap_sort(arr):
    """Convierte el arreglo en un heap y luego lo ordena."""
    n = len(arr)
    # Construcción del heap
    for i in range(n // 2 - 1, -1, -1):
        heapify(arr, n, i)
    # Extrae elementos del heap uno por uno
    for i in range(n - 1, 0, -1):
        arr[i], arr[0] = arr[0], arr[i]
        heapify(arr, i, 0)

# --- Bucket Sort ---
def bucket_sort(arr):
    """Distribuye los elementos en cubetas (buckets), los ordena y combina."""
    if len(arr) == 0:
        return arr
    bucket = [[] for _ in range(10)]
    max_value = max(arr)
    for num in arr:
        index = int(num / (max_value + 1) * 10)
        bucket[index].append(num)
    for i in range(10):
        bucket[i].sort()
    return [num for sublist in bucket for num in sublist]

# --- Radix Sort ---
def radix_sort(arr):
    """Ordena números dígito a dígito (base 10)."""
    RADIX = 10
    maxLength = False
    tmp, placement = -1, 1
    while not maxLength:
        maxLength = True
        buckets = [list() for _ in range(RADIX)]
        for i in arr:
            tmp = i // placement
            buckets[tmp % RADIX].append(i)
            if maxLength and tmp > 0:
                maxLength = False
        a = 0
        for b in range(RADIX):
            buck = buckets[b]
            for i in buck:
                arr[a] = i
                a += 1
        placement *= RADIX
    return arr

# --- Hash Sort (versión simple) ---
def hash_sort(arr):
    """Simula un ordenamiento usando una tabla hash (dict) + sort final."""
    table = {}
    for num in arr:
        table[num] = num
    return sorted(table.keys())

# 2. DEFINICIÓN DE ESCENARIOS DE PRUEBA

# Tamaños de los arreglos
sizes = [100, 1000, 10000, 100000]

# Tipos de orden inicial
types = ["Ordenado", "Medio", "Inverso"]

# Diccionario de algoritmos con sus funciones
algorithms = {
    "Bubble Sort": bubble_sort,
    "Insertion Sort": insertion_sort,
    "Selection Sort": selection_sort,
    "Quick Sort": lambda a: quick_sort(a.copy()),
    "Merge Sort": merge_sort,
    "Heap Sort": heap_sort,
    "Hash Sort": lambda a: hash_sort(a.copy()),
    "Bucket Sort": lambda a: bucket_sort(a.copy()),
    "Radix Sort": lambda a: radix_sort(a.copy())
}

# Lista para almacenar los resultados
results = []

# 3. EJECUCIÓN DE LAS PRUEBAS

# Se recorren los tamaños y tipos para formar los 12 escenarios
for size in sizes:
    base = list(range(size))
    for t in types:
        # Generación del arreglo según el tipo de orden
        if t == "Ordenado":
            arr = base.copy()
        elif t == "Medio":
            arr = base.copy()
            random.shuffle(arr)
        else:  # Inverso
            arr = base[::-1]
        
        # Ejecución de todos los algoritmos sobre el mismo escenario
        for name, func in algorithms.items():
            data = arr.copy()  # copia del arreglo original
            start = time.perf_counter()  # inicio de medición
            func(data)
            end = time.perf_counter()    # fin de medición
            results.append({
                "Método": name,
                "Tamaño": size,
                "Tipo": t,
                "Tiempo (s)": end - start
            })
            print(f"{name:<15} | {size:>6} | {t:<10} | {end - start:.5f}s")

# 4. VISUALIZACIÓN DE TODAS LAS GRÁFICAS POR TAMAÑO

import seaborn as sns
sns.set(style="whitegrid")

df = pd.DataFrame(results)
df["Tiempo (ms)"] = df["Tiempo (s)"] * 1000

# Carpeta de salida para guardar las gráficas
output_dir = "graficas"
os.makedirs(output_dir, exist_ok=True)

# Generamos una gráfica por cada tamaño
for n in [100, 1000, 10000, 100000]:
    df_n = df[df["Tamaño"] == n]

    plt.figure(figsize=(10, 6))
    sns.barplot(
        data=df_n,
        x="Método",
        y="Tiempo (ms)",
        hue="Tipo",
        palette="deep"
    )

    plt.title(f"Tiempos de algoritmos de ordenamiento (n={n})", fontsize=13)
    plt.xlabel("Algoritmo", fontsize=11)
    plt.ylabel("Tiempo (ms)", fontsize=11)
    plt.legend(title="Tipo de arreglo")
    plt.xticks(rotation=30)
    plt.tight_layout()

    # Guardar cada gráfica como imagen
    filename = f"grafico_{n}.png"
    filepath = os.path.join(output_dir, filename)
    plt.savefig(filepath)
    plt.show()

print(f"\n✅ Todas las gráficas se guardaron en la carpeta: {output_dir}")

# 5. TOPS DE RENDIMIENTO

print("\n" + "="*60)
print("TOPS DE ALGORITMOS POR ESCENARIO")
print("="*60)

# Top por tamaño y tipo
for size in sizes:
    for t in types:
        df_subset = df[(df["Tamaño"] == size) & (df["Tipo"] == t)].sort_values(by="Tiempo (s)")
        print(f"\nTamaño={size}, Tipo={t}")
        print(df_subset[["Método", "Tiempo (ms)"]].to_string(index=False))
        mejor = df_subset.iloc[0]["Método"]
        peor = df_subset.iloc[-1]["Método"]
        print(f"Mejor: {mejor} | Peor: {peor}")

# Top global (promedio general)
top_global = df.groupby("Método")["Tiempo (s)"].mean().sort_values()
print("\n" + "="*60)
print("TOP GLOBAL DE ALGORITMOS (Promedio de todos los escenarios)")
print("="*60)
for i, (alg, avg) in enumerate(top_global.items(), start=1):
    print(f"{i:>2}. {alg:<15}  ->  {avg*1000:.4f} ms promedio")